### PR TITLE
[SYCLomatic] Using host_task to implement async_dpct_free

### PR DESCRIPTION
**Pull request by zhiweij1**
Signed-off-by: Jiang, Zhiwei <zhiwei.jiang@intel.com>

### CODE REVIEW

File: clang/runtime/dpct-rt/include/memory.hpp.inc [hunk 1]
```diff
@@ -1195,17 +1192,16 @@ static inline void dpct_free(void *ptr,
 inline void async_dpct_free(std::vector<void *> pointers,
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/memory.hpp.inc]:
```suggestion
inline void async_dpct_free(const std::vector<void *> &pointers,
```

File: clang/runtime/dpct-rt/include/memory.hpp.inc [hunk 2]
```diff
@@ -1195,17 +1192,16 @@ static inline void dpct_free(void *ptr,
...
                             std::vector<sycl::event> events,
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/memory.hpp.inc]:
```suggestion
                            const std::vector<sycl::event> &events,
```

File: clang/runtime/dpct-rt/include/memory.hpp.inc [hunk 3]
```diff
@@ -1195,17 +1192,16 @@ static inline void dpct_free(void *ptr,
...
-         sycl::context ctxt) {
-        sycl::event::wait(events);
-        for (auto p : pointers)
-          sycl::free(p, ctxt);
-      },
-      std::move(pointers), std::move(events), q.get_context());
-  get_current_device().add_task(std::move(t));
+  q.submit([&](sycl::handler &cgh) {
+    cgh.depends_on(events);
+    cgh.host_task([=, &q] {
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/memory.hpp.inc]:
```suggestion
    cgh.host_task([=] {
```

**Review by tomflinda — APPROVED** [review]
LGTM

File: clang/test/dpct/helper_files_ref/include/memory.hpp [hunk 4]
```diff
@@ -800,36 +811,28 @@ static inline void *dpct_malloc(size_t &pitch, size_t x, size_t y,
...
+  detail::dpct_free(ptr, q);
 }
 
-#ifndef DPCT_USM_LEVEL_NONE
 /// Free the device memory pointed by a batch of pointers in \p pointers which
 /// are related to \p q after \p events completed.
 ///
 /// \param pointers The pointers point to the device memory requested to be freed.
 /// \param events The events to be waited.
 /// \param q The sycl::queue the memory relates to.
```
Comment (ziranzha) [inline clang/test/dpct/helper_files_ref/include/memory.hpp:822]: Comments for the function also should be removed

**Review by ziranzha — APPROVED** [review]

**Pull request merged**